### PR TITLE
Bug Fix for Island Edge Candidates

### PR DESCRIPTION
**Pull request by kevinkreiser**
dont use low reachability edge candidates if we ended up rejecting other closer high reachability edge candidates

so we have an issue where we are sometimes using edge candidates that are on disconnected islands and the penalties in thor are not accounting for it (especially for slower modes of travel like pedestrian or bike riders).

heres the before:

broken

the intuition to fixing this is that we shouldnt be using disconnected island edge candidates when we ended up rejecting many other connected edge candidates that were closer to the input location. so the changes in the pr, track the closest of those rejected well-connected edge candidates and then cull any island edges which are further away than those. the result is the following:

fixed

And we get the added benefit that using islands still works:

works

### CODE REVIEW

File: test_requests/island_routes.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+-j '{"locations":[{"lat":-33.4280388615334,"lon":-70.54088473320007},{"lat":-33.4268390099302,"lon":-70.53765535354614}],"costing":"pedestrian"}'
```
Comment (kevinkreiser) [inline test_requests/island_routes.txt]: i couldnt find any good island in utrecht so i made a island routes file instead

File: src/loki/search.cc [hunk 2]
```diff
@@ -697,6 +709,12 @@ struct bin_handler_t {
     // need to go get the actual correlated location with edge_id etc.
     std::unordered_map<Location, PathLocation> searched;
     for (auto& pp : pps) {
+      // remove non-sensical island candidates
+      auto new_end = std::remove_if(pp.unreachable.begin(), pp.unreachable.end(),
+                                    [&pp](const candidate_t& c) -> bool {
+                                      return c.sq_distance > pp.closest_external_reachable;
+                                    });
+      pp.unreachable.erase(new_end, pp.unreachable.end());
```
Comment (kevinkreiser) [inline src/loki/search.cc]: this is where we remove those island candidates that we should have never kept in the first place

File: src/loki/search.cc [hunk 3]
```diff
@@ -642,16 +650,20 @@ struct bin_handler_t {
           // the last one wasnt in the radius so replace it with this one because its better or is
           // in the radius
           if (!last_in_radius) {
+            if (closer_external_reachable)
+              p_itr->closest_external_reachable = batch->back().sq_distance;
```
Comment (kevinkreiser) [inline src/loki/search.cc]: as we are rejecting reachable candidates we keep track of how close the closest one was

File: src/loki/search.cc [hunk 4]
```diff
@@ -642,16 +650,20 @@ struct bin_handler_t {
...
-            // last one is in the radius and this one is not as good so put it on before it
+            // last one and this one are both in the radius but this one is not as good
           } else {
             batch->emplace_back(std::move(*c_itr));
             std::swap(*(batch->end() - 1), *(batch->end() - 2));
           }
-        }
+        } // not in radius or better and reachable and closer than closest one outside of radius
+        else if (closer_external_reachable)
+          p_itr->closest_external_reachable = c_itr->sq_distance;
```
Comment (kevinkreiser) [inline src/loki/search.cc]: same thing, remember how close this rejected one was for later so we can use it to cull low reachability candidates

File: test_requests/island_routes.txt (hunk 1 above)
Comment (danpaz) [reply test_requests/island_routes.txt]: You won't believe it but there is a Prison Island in Utrecht: http://prisonisland.se/en/utrecht/ -- not disconnected from the road network though 😆. 

Anyway I was able to trigger the bug on master with these coordinates in Utrecht with bike costing `5.011605999215277,52.114736949245156;5.020381755701578,52.10586968702347` if that helps.

**Comment by dnesbitt61** [comment]
Curious if this fixes this case: https://github.com/valhalla/valhalla/issues/1575 (auto route). I think I remember looking into this and thinking that edge distance was a factor (there were multiple edge candidates). Strange thing was that this was auto only and not pedestrian and bike that were impacted.

**Comment by kevinkreiser** [comment]
@danpaz that is awesome! if its in our extract i can make a unit test, either way Ill add it to the requests file for good measure!

**Pull request merged**